source/distributed/tria.cc

```diff
@@ -2808,6 +2832,32 @@ namespace parallel
                 && cell->level_subdomain_id() != this->locally_owned_subdomain())
               this->number_cache.level_ghost_owners.insert(cell->level_subdomain_id());
 
+#ifdef DEBUG
+          // Check that level_ghost_owners is symmetric by sending a message
+          // to every one
+          {
+
+            MPI_Barrier(this->mpi_communicator);
+
+            std::vector<MPI_Request> requests;
+            for (std::set<unsigned int>::iterator it = this->number_cache.level_ghost_owners.begin();
+                 it != this->number_cache.level_ghost_owners.end();
+                 ++it)
+              {
+                int dummy = 0;
+                requests.push_back(MPI_Request());
+                MPI_Isend(&dummy, 1, MPI_INT, *it, 9001, this->mpi_communicator, &requests.back());
+                requests.push_back(MPI_Request());
+                MPI_Irecv(&dummy, 1, MPI_INT, *it, 9001, this->mpi_communicator, &requests.back());
```

Comment:
I guess it's that kind of day: I hate to break it to you, pal, but there's a second memory corruption happening here as well. When you do `requests.push_back()`, you may or may not do a memory reallocation. But you've already given the address of the object to the `MPI_Isend/recv` functions which, presumably, could store something into the address. (Or they may not, in case the request object is simply a handle to an opaque object somewhere, in which case the pointer is not stored in the internal machinery of MPI.) In any case, you need to resize the object up front, instead of using `push_back()` and `back()`.